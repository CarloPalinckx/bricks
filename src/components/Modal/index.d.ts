import { Component } from 'react';
import { StyledType } from '../../utility/styled';
declare type PropsType = StyledType & {
    show: boolean;
    title: string;
    size?: 'small' | 'medium' | 'large';
    closeAction?(): void;
    renderFixed?(): JSX.Element;
};
declare class Modal extends Component<PropsType> {
    private styledModalRef;
    private styledModalWrapperRef;
    constructor(props: PropsType);
    handleClickOutside: (event: Event) => void;
    componentDidMount(): void;
    componentWillUnmount(): void;
    render(): JSX.Element;
}
export default Modal;
export { PropsType };
