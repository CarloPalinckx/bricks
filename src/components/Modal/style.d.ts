import _R from 'react';
import { StyledComponentClass as _S } from 'styled-components';
import _T from '../../types/ThemeType';
declare enum ModalSizes {
    small = "480px",
    medium = "792px",
    large = "1068px"
}
declare type ModalThemeType = {
    backgroundColor: string;
    backdropColor: string;
    borderRadius: string;
};
declare type ModalWrapperPropsType = {
    show: boolean;
};
declare const StyledModalWrapper: _S<ModalWrapperPropsType & _R.HTMLProps<HTMLElement>, _T, ModalWrapperPropsType & _R.HTMLProps<HTMLElement>>;
declare type ModalPropsType = {
    modalSize: keyof typeof ModalSizes;
};
declare const StyledModal: _S<ModalPropsType & _R.HTMLProps<HTMLElement>, _T, ModalPropsType & _R.HTMLProps<HTMLElement>>;
export default StyledModal;
export { ModalThemeType, StyledModalWrapper };
