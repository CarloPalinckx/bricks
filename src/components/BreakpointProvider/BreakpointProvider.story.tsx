import { object } from '@storybook/addon-knobs/react';
import { storiesOf } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import BreakpointProvider from '../BreakpointProvider';
import { BreakpointType } from './BreakpointProvider.template';

const StyledDiv = styled.div`
    background: #eee;
    padding: 24px;
`;

storiesOf('BreakpointProvider', module)
    .add('default', () => {
        const breakpoints:BreakpointType = {
            small: 0,
            medium: 300,
            large: 600,
        };

        return (
            <BreakpointProvider breakpoints={object('breakpoints', breakpoints)}>
                {(breakpoint):JSX.Element => {
                    return (
                        <StyledDiv>
                            <div>Current breakpoint is {breakpoint}</div>
                        </StyledDiv>
                    );
                }}
            </BreakpointProvider>
        );
    });
