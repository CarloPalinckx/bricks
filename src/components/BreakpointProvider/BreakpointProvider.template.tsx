import React, { Component } from 'react';

type PropsType = {
    breakpoints:BreakpointType;
    children(breakpoint:keyof BreakpointType):JSX.Element;
};

type StateType = {
    offsetWidth:number;
};

type BreakpointType = {
    large:number;
    small:number;
    medium:number;
};

class BreakpointProvider extends Component<PropsType, StateType> {
    private element:HTMLDivElement;

    public constructor(props:PropsType) {
        super(props);
        this.state = { offsetWidth: 0 };
    }

    private readonly getBreakpoint = (breakpoints:BreakpointType, offsetWidth:number):keyof BreakpointType => Object.keys(this.props.breakpoints)
        .reduce<keyof BreakpointType>(
            (smallestKey, key:keyof BreakpointType):keyof BreakpointType => {
                if (
                    breakpoints[key] > breakpoints[smallestKey] &&
                    offsetWidth >= breakpoints[key]
                ) {
                    return key;
                }

                return smallestKey;
            },
            'small',
        )

    private readonly setBreakpoint = ():void => {
            this.setState({ offsetWidth: this.element.offsetWidth });
    }

    public componentDidMount():void {
        this.setBreakpoint();
        window.addEventListener('resize', this.setBreakpoint);
    }

    public componentWillUnmount():void {
        window.removeEventListener('resize', this.setBreakpoint);
    }

    public render():JSX.Element {
        return (
            <div ref={(element:HTMLDivElement):void => { this.element = element; }}>
                {this.props.children(this.getBreakpoint(this.props.breakpoints, this.state.offsetWidth))}
            </div>
        );
    }
}

export { PropsType, StateType, BreakpointType };
export default BreakpointProvider;
