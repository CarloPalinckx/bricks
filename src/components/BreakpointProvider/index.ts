export { default } from './BreakpointProvider.template';
