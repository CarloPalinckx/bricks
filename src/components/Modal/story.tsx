import { boolean, text, select } from '@storybook/addon-knobs/react';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Modal from '.';
import Button from '../Button';
import ButtonGroup from '../ButtonGroup';
import Text from '../Text';

const demoContent = `
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam ut mi ligula. Phasellus tellus nulla,
cursus sit amet dolor eu, sodales facilisis tortor. Maecenas sed arcu quis est pharetra molestie sed
eu leo. Mauris consequat mauris et eros gravida vestibulum. Phasellus convallis ipsum quis nisl lacinia,
a pulvinar est porta. Nunc tempus vulputate dapibus. In eget venenatis orci. Pellentesque habitant morbi
tristique senectus et netus et malesuada fames ac turpis egestas. Maecenas finibus lorem et quam imperdiet,
id vestibulum odio molestie. Curabitur euismod sit amet tortor et imperdiet. Nam a nisl quis lorem porta
pharetra. Duis sed magna vel odio ullamcorper gravida eu et nibh. Mauris tempor libero purus, vitae lacinia
felis hendrerit in. Vestibulum rhoncus vitae ipsum vel placerat.
eu leo. Mauris consequat mauris et eros gravida vestibulum. Phasellus convallis ipsum quis nisl lacinia,
a pulvinar est porta. Nunc tempus vulputate dapibus. In eget venenatis orci. Pellentesque habitant morbi
tristique senectus et netus et malesuada fames ac turpis egestas. Maecenas finibus lorem et quam imperdiet,
id vestibulum odio molestie. Curabitur euismod sit amet tortor et imperdiet. Nam a nisl quis lorem porta
pharetra. Duis sed magna vel odio ullamcorper gravida eu et nibh. Mauris tempor libero purus, vitae lacinia
felis hendrerit in. Vestibulum rhoncus vitae ipsum vel placerat.
eu leo. Mauris consequat mauris et eros gravida vestibulum. Phasellus convallis ipsum quis nisl lacinia,
a pulvinar est porta. Nunc tempus vulputate dapibus. In eget venenatis orci. Pellentesque habitant morbi
tristique senectus et netus et malesuada fames ac turpis egestas. Maecenas finibus lorem et quam imperdiet,
id vestibulum odio molestie. Curabitur euismod sit amet tortor et imperdiet. Nam a nisl quis lorem porta
pharetra. Duis sed magna vel odio ullamcorper gravida eu et nibh.`;

storiesOf('Modal', module)
    .add('Default', () => {
        return (
            <Modal
                show={boolean('show', true)}
                size={select('size', ['small', 'medium', 'large'], 'large')}
                title="Would you like me to be your role modal?"
                closeAction={(): boolean => confirm('You are now closing this modal, do you wish to continue?')}
                renderFixed={(): JSX.Element => (
                    <ButtonGroup>
                        <Button variant="primary" title="Activate" />
                        <Button variant="plain" title="Close" />
                    </ButtonGroup>
                )}
            >
                <Text>{text('contents', demoContent)}</Text>
            </Modal>
        );
    })
    .add('Without closeAction', () => {
        return (
            <Modal
                show={boolean('show', true)}
                size={select('size', ['small', 'medium', 'large'], 'large')}
                title="Would you like me to be your role modal?"
                renderFixed={(): JSX.Element => (
                    <ButtonGroup>
                        <Button variant="primary" title="Activate" />
                        <Button variant="plain" title="Close" />
                    </ButtonGroup>
                )}
            >
                <Text>{text('contents', demoContent)}</Text>
            </Modal>
        );
    });
