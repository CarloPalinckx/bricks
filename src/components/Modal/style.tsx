import _R from 'react';
import { StyledComponentClass as _S } from 'styled-components';
import _T from '../../types/ThemeType';
import styled, { withProps } from '../../utility/styled';

type ModalThemeType = {
    backgroundColor: string;
    backdropColor: string;
    borderRadius: string;
};

type ModalWrapperPropsType = {
    show: boolean;
};

const StyledModalWrapper = withProps<ModalWrapperPropsType>(styled.div)`
    transition: opacity 100ms;
    background: ${({ theme }): string => theme.Modal.backdropColor};
    position: fixed;
    height: 100%;
    width: 100%;
    top: 0;
    left: 0;
    box-sizing: border-box;
    display: flex;
    justify-content: center;
    align-items: center;

    ${({ show }): string => (show ? 'opacity: 1;' : 'opacity: 0; pointer-events: none;')};
`;

const StyledModal = styled.div`
    margin: auto;
    max-width: 600px;
    width: calc(100% - 24px);
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    min-height: 300px;
    min-width: 150px;
    height: calc(100% - 24px);
    overflow: hidden;
    max-height: calc(300px + (600 - 300) * (100vh - 300px) / (900 - 300));
    background: ${({ theme }): string => theme.Modal.backgroundColor};
    border-radius: ${({ theme }): string => theme.Modal.borderRadius};
`;

export default StyledModal;
export { ModalThemeType, StyledModalWrapper };
