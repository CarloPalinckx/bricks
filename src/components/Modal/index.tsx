import React, { Component } from 'react';
import { StyledType } from '../../utility/styled';
import trbl from '../../utility/trbl';
import Box from '../Box';
import BreakpointProvider from '../BreakpointProvider';
import Button from '../Button';
import Contrast from '../Contrast';
import Heading from '../Heading';
import Icon from '../Icon';
import ScrollBox from '../ScrollBox';
import TransitionAnimation from '../TransitionAnimation';
import StyledModal, { StyledModalWrapper } from './style';

type PropsType = StyledType & {
    show: boolean;
    title: string;
    size?: 'small' | 'medium' | 'large';
    closeAction?(): void;
    renderFixed?(): JSX.Element;
};

class Modal extends Component<PropsType> {
    private styledModalRef: HTMLDivElement;
    private styledModalWrapperRef: HTMLDivElement;

    public constructor(props: PropsType) {
        super(props);
    }

    public handleClickOutside = (event: Event): void => {
        if (
            this.props.closeAction !== undefined &&
            this.props.show &&
            this.styledModalWrapperRef.contains(event.target as Node) &&
            !this.styledModalRef.contains(event.target as Node)
        ) {
            this.props.closeAction();
        }
    };

    public componentDidMount(): void {
        document.addEventListener('mousedown', this.handleClickOutside, false);
    }

    public componentWillUnmount(): void {
        document.removeEventListener('mousedown', this.handleClickOutside, false);
    }

    public render(): JSX.Element {
        return (
            <StyledModalWrapper
                show={this.props.show}
                innerRef={(ref): void => {
                    this.styledModalWrapperRef = ref;
                }}
            >
                <TransitionAnimation key={0} show={this.props.show} animation="zoom">
                    <BreakpointProvider breakpoints={{ small: 0, medium: 320, large: 1200 }}>
                        {(breakpoint): JSX.Element => (
                            <StyledModal
                                modalSize={this.props.size !== undefined ? this.props.size : 'large'}
                                innerRef={(ref): void => {
                                    this.styledModalRef = ref;
                                }}
                            >
                                <Box
                                    shrink={0}
                                    margin={breakpoint === 'small' ? trbl(24) : trbl(24, 36)}
                                    alignItems="flex-start"
                                    alignContent="center"
                                    justifyContent="space-between"
                                >
                                    <Heading hierarchy={2}>{this.props.title}</Heading>
                                    {this.props.closeAction !== undefined && (
                                        <Box
                                            margin={trbl(-12, -12, -6, 0)}
                                            alignContent="center"
                                            justifyContent="flex-end"
                                            alignItems="center"
                                            grow={0}
                                        >
                                            <Button.Flat
                                                variant="primary"
                                                title="close"
                                                onClick={this.props.closeAction}
                                            >
                                                <Icon size="small" icon="close" />
                                            </Button.Flat>
                                        </Box>
                                    )}
                                </Box>
                                <ScrollBox>
                                    <Box padding={breakpoint === 'small' ? trbl(0, 24, 24, 24) : trbl(0, 36, 36, 36)}>
                                        {this.props.children}
                                    </Box>
                                </ScrollBox>
                                {this.props.renderFixed && (
                                    <Contrast>
                                        <Box direction="column" alignItems="stretch" shrink={0} padding={trbl(24, 36)}>
                                            {this.props.renderFixed()}
                                        </Box>
                                    </Contrast>
                                )}
                            </StyledModal>
                        )}
                    </BreakpointProvider>
                </TransitionAnimation>
            </StyledModalWrapper>
        );
    }
}

export default Modal;
export { PropsType };
