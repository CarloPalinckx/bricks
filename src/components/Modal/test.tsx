import toJson from 'enzyme-to-json';
import React from 'react';
import Modal from '.';
import { mountWithTheme, shallowWithTheme } from '../../utility/styled';
import BreakpointProvider from '../BreakpointProvider';
import { PropsType } from '../BreakpointProvider/';
import StyledModal, { StyledModalWrapper } from './style';

jest.mock('../ScrollBox', () => jest.fn().mockImplementation((props: PropsType): string => 'div'));

jest.mock('../BreakpointProvider', () =>
    jest.fn().mockImplementation((props: PropsType): JSX.Element => {
        return props.children('large');
    }),
);

describe('Modal', () => {
    it('should render with a small breakpoint', () => {
        (BreakpointProvider as jest.Mock<BreakpointProvider>).mockImplementationOnce(
            (props: PropsType): JSX.Element => {
                return props.children('small');
            },
        );

        const component = mountWithTheme(
            <Modal show={true} renderFixed={(): JSX.Element => <div>bar</div>} title="Foo" />,
        );

        expect(component.find(StyledModal).length).toBe(1);
    });

    it('should render with a medium breakpoint', () => {
        (BreakpointProvider as jest.Mock<BreakpointProvider>).mockImplementationOnce(
            (props: PropsType): JSX.Element => {
                return props.children('medium');
            },
        );

        const component = mountWithTheme(<Modal show={true} title="Foo" />);

        expect(component.find(StyledModal).length).toBe(1);
    });

    it('should be possible to close the modal view', () => {
        const clickMock = jest.fn();
        const component = shallowWithTheme(<Modal show={true} title="Foo" closeAction={clickMock} />);
        const styledModalWrapper = component.find(StyledModalWrapper);

        styledModalWrapper.simulate('click');

        expect(toJson(component.dive())).toMatchSnapshot();
        expect(clickMock).toHaveBeenCalled();

        const componentNoAction = shallowWithTheme(<Modal show={true} title="Foo" closeAction={undefined} />);
        componentNoAction.simulate('click');
    });

    it('should render closed', () => {
        const component = mountWithTheme(<Modal show={false} title="Foo" />);
        expect(component.length).toBe(1);
    });
});
