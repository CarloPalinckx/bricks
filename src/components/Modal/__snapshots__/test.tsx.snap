// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should be possible to close the modal view 1`] = `
.c0 {
  -webkit-transition: opacity 100ms;
  transition: opacity 100ms;
  background: rgba(0,0,0,0.3);
  position: fixed;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 1;
}

<div
  className="c0"
  onClick={[Function]}
>
  <TransitionAnimation
    animation="zoom"
    key="0"
    show={true}
  >
    <mockConstructor
      breakpoints={
        Object {
          "large": 1200,
          "medium": 320,
          "small": 0,
        }
      }
    />
  </TransitionAnimation>
</div>
`;
